File: experimental/ImPlay/Implay.Core/Services/SubtitleParser.cs
using System.Text.RegularExpressions;

namespace ImPlay.Core.Services;

/// <summary>A single timed subtitle entry.</summary>
public sealed record SubtitleLine(TimeSpan Start, TimeSpan End, string Text);

/// <summary>
/// Minimal parser for SRT and basic ASS/SSA subtitle files.
/// Only the timing and visible text is extracted — styling is ignored.
/// </summary>
public static class SubtitleParser
{
    // ── SRT ───────────────────────────────────────────────────────────────────
    // Matches: 00:01:23,456 --> 00:01:27,890
    private static readonly Regex SrtTimecode = new(
        @"(\d+):(\d+):(\d+)[,\.](\d+)\s*-->\s*(\d+):(\d+):(\d+)[,\.](\d+)",
        RegexOptions.Compiled);

    // ── ASS / SSA ─────────────────────────────────────────────────────────────
    // Matches the Dialogue line: Layer,Start,End,Style,Name,MarginL,R,V,Effect,Text
    private static readonly Regex AssDialogue = new(
        @"^Dialogue:\s*\d+,(\d+:\d+:\d+\.\d+),(\d+:\d+:\d+\.\d+),[^,]*,[^,]*,[^,]*,[^,]*,[^,]*,[^,]*,(.*)$",
        RegexOptions.Compiled);

    // Strips ASS inline override tags like {\pos(...)}, {\an8}, {\c&H...} etc.
    private static readonly Regex AssInlineTags = new(@"\{[^}]*\}", RegexOptions.Compiled);

    // Strips HTML-like tags left in some SRT files
    private static readonly Regex HtmlTags = new(@"<[^>]+>", RegexOptions.Compiled);

    /// <summary>Parse a subtitle file, auto-detecting SRT vs ASS/SSA.</summary>
    public static List<SubtitleLine> Parse(string filePath)
    {
        if (!File.Exists(filePath)) return [];

        string[] lines;
        try { lines = File.ReadAllLines(filePath); }
        catch { return []; }

        var ext = Path.GetExtension(filePath).ToLowerInvariant();
        return ext is ".ass" or ".ssa" ? ParseAss(lines) : ParseSrt(lines);
    }

    // ── SRT parser ────────────────────────────────────────────────────────────

    private static List<SubtitleLine> ParseSrt(string[] lines)
    {
        var result = new List<SubtitleLine>();
        int i = 0;

        while (i < lines.Length)
        {
            // Skip blank lines and sequence numbers
            while (i < lines.Length && string.IsNullOrWhiteSpace(lines[i])) i++;
            if (i >= lines.Length) break;

            // Skip a purely numeric sequence number if present
            if (int.TryParse(lines[i].Trim(), out _)) i++;
            if (i >= lines.Length) break;

            // Timecode line
            var m = SrtTimecode.Match(lines[i]);
            if (!m.Success) { i++; continue; }

            var start = ParseSrtTime(m, 1);
            var end   = ParseSrtTime(m, 5);
            i++;

            // Text lines until blank or EOF
            var text = new System.Text.StringBuilder();
            while (i < lines.Length && !string.IsNullOrWhiteSpace(lines[i]))
            {
                if (text.Length > 0) text.Append('\n');
                text.Append(HtmlTags.Replace(lines[i], ""));
                i++;
            }

            var t = text.ToString().Trim();
            if (t.Length > 0)
                result.Add(new SubtitleLine(start, end, t));
        }

        return result;
    }

    private static TimeSpan ParseSrtTime(Match m, int g) =>
        new(0,
            int.Parse(m.Groups[g].Value),
            int.Parse(m.Groups[g + 1].Value),
            int.Parse(m.Groups[g + 2].Value),
            int.Parse(m.Groups[g + 3].Value));

    // ── ASS / SSA parser ──────────────────────────────────────────────────────

    private static List<SubtitleLine> ParseAss(string[] lines)
    {
        var result = new List<SubtitleLine>();

        foreach (var line in lines)
        {
            var m = AssDialogue.Match(line);
            if (!m.Success) continue;

            if (!TryParseAssTime(m.Groups[1].Value, out var start)) continue;
            if (!TryParseAssTime(m.Groups[2].Value, out var end))   continue;

            var raw  = m.Groups[3].Value;
            // Strip override tags and \N / \n hard line breaks
            var text = AssInlineTags.Replace(raw, "")
                          .Replace(@"\N", "\n")
                          .Replace(@"\n", "\n")
                          .Trim();

            if (text.Length > 0)
                result.Add(new SubtitleLine(start, end, text));
        }

        return result;
    }

    // ASS time format: H:MM:SS.cc  (centiseconds)
    private static bool TryParseAssTime(string s, out TimeSpan t)
    {
        t = default;
        var parts = s.Split(':');
        if (parts.Length != 3) return false;
        if (!int.TryParse(parts[0], out var h)) return false;
        if (!int.TryParse(parts[1], out var min)) return false;
        var secParts = parts[2].Split('.');
        if (!int.TryParse(secParts[0], out var sec)) return false;
        var cs = secParts.Length > 1 && int.TryParse(secParts[1], out var x) ? x * 10 : 0;
        t = new TimeSpan(0, h, min, sec, cs);
        return true;
    }
}

File: experimental/implayfsharpavalonia/Core/SubtitleParser.fs
namespace ImPlay.Core.Services

open System
open System.IO
open System.Text
open System.Text.RegularExpressions
open System.Collections.Generic
open ImPlay.Core.Models

type SubtitleLine = { Start : TimeSpan; End : TimeSpan; Text : string }

module SubtitleParser =
    // Matches: 00:01:23,456 --> 00:01:27,890
    let srtTimecode = Regex(@"(\d+):(\d+):(\d+)[,\.](\d+)\s*-->\s*(\d+):(\d+):(\d+)[,\.](\d+)", RegexOptions.Compiled)
    // Matches the Dialogue line: Layer,Start,End,Style,Name,MarginL,R,V,Effect,Text
    let assDialogue = Regex(@"^Dialogue:\s*\d+,(\d+:\d+:\d+\.\d+),(\d+:\d+:\d+\.\d+),[^,]*,[^,]*,[^,]*,[^,]*,[^,]*,[^,]*,(.*)$", RegexOptions.Compiled)
    // Strips ASS inline override tags
    let assInlineTags = Regex(@"\{[^}]*\}", RegexOptions.Compiled)
    // Strips HTML-like tags
    let htmlTags = Regex(@"<[^>]+>", RegexOptions.Compiled)

    let private parseSrtTime (m: Match) (g: int) =
        TimeSpan(0, 
            int m.Groups.[g].Value, 
            int m.Groups.[g + 1].Value, 
            int m.Groups.[g + 2].Value, 
            int m.Groups.[g + 3].Value)

    let private tryParseAssTime (s: string) =
        let parts = s.Split(':')
        if parts.Length <> 3 then None
        else
            match Int32.TryParse(parts.[0]), Int32.TryParse(parts.[1]) with
            | (true, h), (true, m) ->
                let secParts = parts.[2].Split('.')
                match Int32.TryParse(secParts.[0]) with
                | (true, s) ->
                    let cs = if secParts.Length > 1 then match Int32.TryParse(secParts.[1]) with (true, x) -> x * 10 | _ -> 0 else 0
                    Some (TimeSpan(0, h, m, s, cs))
                | _ -> None
            | _ -> None

    let private parseSrt (lines: string[]) =
        let result = ResizeArray<SubtitleLine>()
        let mutable i = 0
        while i < lines.Length do
            while i < lines.Length && String.IsNullOrWhiteSpace(lines.[i]) do i <- i + 1
            if i < lines.Length then
                if fst (Int32.TryParse(lines.[i].Trim())) then i <- i + 1
                if i < lines.Length then
                    let m = srtTimecode.Match(lines.[i])
                    if not m.Success then i <- i + 1
                    else
                        let start = parseSrtTime m 1
                        let endPos = parseSrtTime m 5
                        i <- i + 1
                        let sb = StringBuilder()
                        while i < lines.Length && not (String.IsNullOrWhiteSpace(lines.[i])) do
                            if sb.Length > 0 then sb.Append('\n') |> ignore
                            sb.Append(htmlTags.Replace(lines.[i], "")) |> ignore
                            i <- i + 1
                        let t = sb.ToString().Trim()
                        if t.Length > 0 then
                            result.Add({ Start = start; End = endPos; Text = t })
        result |> Seq.toList

    let private parseAss (lines: string[]) =
        let result = ResizeArray<SubtitleLine>()
        for line in lines do
            let m = assDialogue.Match(line)
            if m.Success then
                match tryParseAssTime m.Groups.[1].Value, tryParseAssTime m.Groups.[2].Value with
                | Some start, Some endPos ->
                    let raw = m.Groups.[3].Value
                    let t = assInlineTags.Replace(raw, "").Replace(@"\N", "\n").Replace(@"\n", "\n").Trim()
                    if t.Length > 0 then
                        result.Add({ Start = start; End = endPos; Text = t })
                | _ -> ()
        result |> Seq.toList

    let parse (filePath: string) =
        if not (File.Exists(filePath)) then []
        else
            try
                let lines = File.ReadAllLines(filePath)
                let ext = Path.GetExtension(filePath).ToLowerInvariant()
                if ext = ".ass" || ext = ".ssa" then parseAss lines
                else parseSrt lines
            with _ -> []
